Спринт 15

cyberschoolinmoscow opened the pull request:
1) Код разделён на три слоя: Data, Domain и Presentation.
2) Все данные, передаваемые между слоями Data и Domain, проходят через DTO.
3) Доступ к данным осуществляется через репозитории, которые инкапсулируют логику получения и сохранения данных.
4) Бизнес-логика вынесена в интеракторы, которые взаимодействуют с репозиториями.
Работа с интеракторами организована через интерфейсы.
5) Экземпляры интеракторов в Activity, необходимые для взаимодействия с Domain-слоем, создаются при помощи Creator-класса.

cyberschoolinmoscow committed:
Доавлено:
1) Разделение по пакетам data, domain, presentation
2) Созданы интерфейсы репозиториев и их реализации для осуществления поиска треков
3) Созданы интерфейсы интеракторов и их реализации для осуществления поиска треков
4) Создан класс Creator для инициализации зависимостей, необходимыз для поиска трека
5) Для SearchActivity реализовано взаимодействие со слоем domain через интерактор, созданный при помощи класса Creator

cyberschoolinmoscow committed:
Доавлено:
1) Интерактор для SharedPreferenceRepository, созданный при помощи класса Creator, добавлен в SettingsActivity

cyberschoolinmoscow committed:
Доавлено:
1) TrackManager и SettingsManager добавлены

cyberschoolinmoscow committed:
Удален:
1) TrackPreferences.kt

cyberschoolinmoscow committed:
1) В классе SearchActivity ключи для передачи данных вынесены в константу.
2) В классе SharedPreferenceRepository параметры передаются в конструктор через private val.
3) Из класса SettingsInteractorImpl убрана зависимость от SettingsManager.
4) В класса TracksRepositoryImpl из сети получается модель трека TrackDto и выполняется маппинг TrackDto в Track.
5) Класс TracksSearchResponse удален.
6) В классе App записана ссылка на Application и используется везде, где нужен контекст.
7) Удален весь закоментированный код.
8) Исправлена проблема:
если включить темную тему в настройках, то при каждом заходе на экран настроек свитч "Тёмная тема" будет в неактивном состоянии и чтобы выключить темную тему нужно два раза на него нажать
9) Исправлена проблема:
все данные, передаваемые между слоями Data и Domain, должны проходить через DTO - TrackDTO не используется, используется одна и та же модель Track для работы с треками во всех слоях
10) Исправлена проблема:
Код должен быть хорошо структурирован, без избыточных зависимостей между слоями - Имеется избыточная зависимость SettingsInteractorImpl от SettingsManager из data слоя. Интерактор может зависеть только от репозиториев

cyberschoolinmoscow committed:
Первая попытка

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/App.kt:
```diff
@@ -1,17 +1,21 @@
     override fun onCreate() {
         super.onCreate()
-        sharedPreferences = getSharedPreferences("MODE", Context.MODE_PRIVATE)
+//        sharedPreferences = getSharedPreferences("MODE", Context.MODE_PRIVATE)
```
Закоментированный код лучше сразу удалять. Хранить закоментированный код в проекте - плохая практика, в случае чего этот код можно восстановить с помощью истории гита

ArturNurtdinov replied:
Лучше пройтись по всем другим файлам и удалить весь закоментированный код

cyberschoolinmoscow replied:
Спасибо! Исправила.

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/creator/Creator.kt:
```diff
@@ -0,0 +1,44 @@
+import com.practicum.playlistmaker.domain.impl.TracksInteractorImpl
+
+object Creator {
+    private fun getTracksRepository(context: Context): TracksRepository {
```
В этом объекте можно не передавать в каждой функции Context в качестве аргумента, а можно в классе App.onCreate записать ссылку на `Application` в этом объекте и использовать её везде, где нужен контекст

cyberschoolinmoscow replied:
Спасибо! Исправила.

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/data/dto/TracksSearchResponse.kt:
```diff
@@ -0,0 +1,7 @@
+package com.practicum.playlistmaker.data.dto
+
+class TracksSearchResponse(
```
Класс не используется

cyberschoolinmoscow replied:
> Класс не используется


Спасибо! Исправила.

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/data/network/TracksRepositoryImpl.kt:
```diff
@@ -0,0 +1,51 @@
+        if (response.resultCode == 200) {
+            return (response as TrackResponse).results.map {
+//                Track(it.trackName, it.artistName, it.trackTime, it.artworkUrl100, it.collectionName,it.releaseDate,it.primaryGenreName,it.country,it.previewUrl)  }
+                Track(
```
Здесь сейчас происходит маппинг `Track` -> `Track`, что бессмысленно, т.к. это просто создание тех же объектов. С сети нужно получать модель трека `TrackDto` и потом мапить `TrackDto` -> `Track`

cyberschoolinmoscow replied:
> Здесь сейчас происходит маппинг `Track` -> `Track`, что бессмысленно, т.к. это просто создание тех же объектов. С сети нужно получать модель трека `TrackDto` и потом мапить `TrackDto` -> `Track`


Спасибо! Исправила.

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/domain/impl/SettingsInteractorImpl.kt:
```diff
@@ -0,0 +1,39 @@
+package com.practicum.playlistmaker.domain.impl
+
+import android.content.Context
+import com.practicum.playlistmaker.data.preferences.SettingsManager
```
Домейн классы не должны зависеть от классов из data слоя, здесь нужно убрать зависимость от `SettingsManager` и использовать только репозиторий

cyberschoolinmoscow replied:
> Домейн классы не должны зависеть от классов из data слоя, здесь нужно убрать зависимость от `SettingsManager` и использовать только репозиторий


Спасибо! Исправила.

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/domain/impl/SettingsInteractorImpl.kt:
```diff
@@ -0,0 +1,39 @@
+import com.practicum.playlistmaker.domain.api.SharedPreferenceRepository
+
+class SettingsInteractorImpl(
+    var repository: SharedPreferenceRepository
```
```suggestion
    private val repository: SharedPreferenceRepository
```
Лучше именно через private val передавать в конструктор параметры, т.к. иначе эти классы станут доступны снаружи интерактора, что обесценивает разделение кода на слои

cyberschoolinmoscow replied:
> Лучше именно через private val передавать в конструктор параметры, т.к. иначе эти классы станут доступны снаружи интерактора, что обесценивает разделение кода на слои


Спасибо! Исправила.

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/presentation/tracks/SearchActivity.kt:
```diff
@@ -0,0 +1,249 @@
+                    this@SearchActivity.startActivity(
+                        Intent(
+                            this@SearchActivity, AudioPlayerActivity::class.java
+                        ).putExtra("track", track.serializeTrack())
```
Ключи для передачи данных лучше вынести в константу, т.к. они используются в более чем 1 месте. Ещё один плюс вынесения в константы - в будущем легче будет этот экран переиспользовать

cyberschoolinmoscow replied:
> Ключи для передачи данных лучше вынести в константу, т.к. они используются в более чем 1 месте. Ещё один плюс вынесения в константы - в будущем легче будет этот экран переиспользовать

Спасибо! Исправила.

cyberschoolinmoscow committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	app/src/main/java/com/practicum/playlistmaker/creator/Creator.kt
#	app/src/main/java/com/practicum/playlistmaker/data/SharedPreferenceRepositoryImp.kt
#	app/src/main/java/com/practicum/playlistmaker/data/preferences/SettingsManager.kt
#	app/src/main/java/com/practicum/playlistmaker/domain/api/SettingsInteractor.kt
#	app/src/main/java/com/practicum/playlistmaker/domain/impl/SettingsInteractorImpl.kt
#	app/src/main/java/com/practicum/playlistmaker/presentation/audioplayer/AudioPlayerActivity.kt
#	app/src/main/java/com/practicum/playlistmaker/presentation/main/MainActivity.kt
#	app/src/main/java/com/practicum/playlistmaker/presentation/settings/SettingsActivity.kt
#	app/src/main/java/com/practicum/playlistmaker/presentation/tracks/SearchActivity.kt

cyberschoolinmoscow committed:
1) В классе SearchActivity ключи для передачи данных вынесены в константу.
2) В классе SharedPreferenceRepository параметры передаются в конструктор через private val.
3) Из класса SettingsInteractorImpl убрана зависимость от SettingsManager.
4) В класса TracksRepositoryImpl из сети получается модель трека TrackDto и выполняется маппинг TrackDto в Track.
5) Класс TracksSearchResponse удален.
6) В классе App записана ссылка на Application и используется везде, где нужен контекст.
7) Удален весь закоментированный код.
8) Исправлена проблема:
если включить темную тему в настройках, то при каждом заходе на экран настроек свитч "Тёмная тема" будет в неактивном состоянии и чтобы выключить темную тему нужно два раза на него нажать
9) Исправлена проблема:
все данные, передаваемые между слоями Data и Domain, должны проходить через DTO - TrackDTO не используется, используется одна и та же модель Track для работы с треками во всех слоях
10) Исправлена проблема:
Код должен быть хорошо структурирован, без избыточных зависимостей между слоями - Имеется избыточная зависимость SettingsInteractorImpl от SettingsManager из data слоя. Интерактор может зависеть только от репозиториев

cyberschoolinmoscow commented:
1) В классе SearchActivity ключи для передачи данных вынесены в константу.
2) В классе SharedPreferenceRepository параметры передаются в конструктор через private val.
3) Из класса SettingsInteractorImpl убрана зависимость от SettingsManager.
4) В класса TracksRepositoryImpl из сети получается модель трека TrackDto и выполняется маппинг TrackDto в Track.
5) Класс TracksSearchResponse удален.
6) В классе App записана ссылка на Application и используется везде, где нужен контекст.
7) Удален весь закоментированный код.
8) Исправлена проблема: 
если включить темную тему в настройках, то при каждом заходе на экран настроек свитч "Тёмная тема" будет в неактивном состоянии и чтобы выключить темную тему нужно два раза на него нажать
9) Исправлена проблема: 
все данные, передаваемые между слоями Data и Domain, должны проходить через DTO - TrackDTO не используется, используется одна и та же модель Track для работы с треками во всех слоях
10) Исправлена проблема:
Код должен быть хорошо структурирован, без избыточных зависимостей между слоями - Имеется избыточная зависимость SettingsInteractorImpl от SettingsManager из data слоя. Интерактор может зависеть только от репозиториев

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/presentation/settings/SettingsActivity.kt:
Все криты поправлены, молодец. Заметил тут, что на экране настроек в тёмной теме цвет шапки экрана отличается от цвета бэкраунда, думаю там одной строчкой это можно поправить, поэтому не стал отдельным критом сейчас писать, поправь пожалуйста к следующему спринту
